File: tests/function.py
# Function and Call 函数定义和调用

## 位置参数
def mul(a, b):
    """simple multiplication"""
    return a * b

assert mul(10, 2) == 20
assert mul(1, b=10) == 10
assert mul(a=2, b=4) == 8
assert mul(a=3, b=5) == 15


## 默认值参数
def sub(a, b=2):
    """minus function"""
    return a - b

assert sub(10, 10) == 0
assert sub(10) == 8
assert sub(10, b=8) == 2
assert sub(a=11, b=3) == 8
assert sub(b=11, a=3) == -8


## vararg，在 python 中如 *args 的写法，写在位置参数的最后，分隔位置参数和键值参数
## 多余的位置参数都由它来吸收
def sum(first, second, *more):
    """sum of the parameters"""
    s = first + second
    for i in more:
        s += i
    return s

assert sum(1, 2, 3, 4) == 10
assert sum(1, 2, 3) == 6
assert sum(1, 2) == 3

assert sum(1, second=2) == 3
assert sum(first=1, second=2) == 3



## 键值参数，在定义 vararg 之后的都是键值参数
## 有默认值为可选值，无默认值为必选值
def add(a, b, *c, d, e):
    return a + b + d + e

assert add(1, 2, 3, 4, d=1, e=2) == 6
assert add(1, 2, d=1, e=2) == 6
assert add(1, b=2, d=1, e=2) == 6
assert add(a=1, b=2, d=1, e=2) == 6
assert add(d=1, e=2, a=1, b=2) == 6

def add(a, b, *c, d=0, e):
    return a + b + d + e

assert add(1, 2, 3, 4, d=1, e=2) == 6
assert add(1, 2, 3, 4, e=2) == 5
assert add(1, 2, d=1, e=2) == 6
assert add(1, 2, e=2) == 5
assert add(1, b=2, d=1, e=2) == 6
assert add(1, b=2, e=2) == 5
assert add(a=1, b=2, d=1, e=2) == 6
assert add(a=1, b=2, e=2) == 5
assert add(d=1, e=2, a=1, b=2) == 6
assert add(e=2, a=1, b=2) == 5


## kwargs，在定义所有键值之后，可定义 kwargs，吸收所有未定义的键值对
## 键值的调用可以复用 c 和 f
def add(a, b, *c, d, e, **f):
    return a + b + d + e + f['c'] + f['f']

assert add(1, 2, 3, 4, d=1, e=2, c=3, f=4) == 13
assert add(1, 2, d=1, e=2, c=3, f=4) == 13
assert add(1, b=2, d=1, e=2, c=3, f=4) == 13
assert add(a=1, b=2, d=1, e=2, c=3, f=4) == 13
assert add(d=1, e=2, a=1, b=2, c=3, f=4) == 13


## 函数作为参数
## 在 python 中，函数作为一等公民，可以作为其它函数的参数
def registerClickEvent(callback):
    """register a function as callback"""
    event = { "id": "click" }
    callback(event)

def callback(event):
    """simple callback"""
    assert event["id"] == "click"

## 调用
registerClickEvent(callback)


## 嵌套函数定义，并返回一个函数
## 函数作为一等公民，也可以作为返回值
def plus_factory(plus_num):
    """plus factory"""
    def plus_func(n):
        """plus function"""
        return n + plus_num
    return plus_func

## 调用
plus_10 = plus_factory(10)
assert plus_10(20) == 30
assert plus_10(-10) == 0


## 装饰器
def clamp_wrapper(func):
    """a clamp decorator"""
    def wrapper(*args, **kwargs):
        res = func(*args, **kwargs)
        if res < 0:
            res = 0
        return res
    return wrapper

@clamp_wrapper
def minus(a, b=10, *c, d=10, e=1, **f):
    return a - d

assert minus(10, 1) == 0
assert minus(10, 20) == 0
assert minus(1) == 0
assert minus(1, d=0) == 1


## 多层装饰器
def clamp_wrapper_with(left, right):
    """a clamp decorator with left and right limit"""
    def clamp_wrapper(func):
        def wrapper(*args, **kwargs):
            res = func(*args, **kwargs)
            if res < left:
                res = left
            elif res > right:
                res = right
            return res
        return wrapper
    return clamp_wrapper
        
@clamp_wrapper_with(1, 10)
def sum(a, b, *c, d, e, **f):
    t = 0
    for i in c:
        t += i
    return t + a + b + d + e

## 调用
assert sum(1, 2, 3, d=4, e=5) == 10
assert sum(-1, -2, d=-3, e=-4) == 1
assert sum(a=1, b=-1, c=0, d=1, e=-10) == 1

File: tests/function.lua
local function mul(kvs, a, b)
    a = get_posarg(kvs, 'a', a, nil, 'mul')
    b = get_posarg(kvs, 'b', b, nil, 'mul')
    --[[simple multiplication]]
    return (a * b)
end
assert(bool((mul(merge_kwargs({}, {}), 10, 2) == 20)))
assert(bool((mul(merge_kwargs({b = 10}, {}), 1) == 10)))
assert(bool((mul(merge_kwargs({a = 2, b = 4}, {})) == 8)))
assert(bool((mul(merge_kwargs({a = 3, b = 5}, {})) == 15)))
local function sub(kvs, a, b)
    a = get_posarg(kvs, 'a', a, nil, 'sub')
    b = get_posarg(kvs, 'b', b, 2, 'sub')
    --[[minus function]]
    return (a - b)
end
assert(bool((sub(merge_kwargs({}, {}), 10, 10) == 0)))
assert(bool((sub(merge_kwargs({}, {}), 10) == 8)))
assert(bool((sub(merge_kwargs({b = 8}, {}), 10) == 2)))
assert(bool((sub(merge_kwargs({a = 11, b = 3}, {})) == 8)))
assert(bool((sub(merge_kwargs({b = 11, a = 3}, {})) == (-8))))
local function sum(kvs, first, second, ...)
    first = get_posarg(kvs, 'first', first, nil, 'sum')
    second = get_posarg(kvs, 'second', second, nil, 'sum')
    local more = list {...}
    --[[sum of the parameters]]
    local s = (first + second)
    for _, i in more do
        s = (s + i)
        ::loop_label_1::
    end
    return s
end
assert(bool((sum(merge_kwargs({}, {}), 1, 2, 3, 4) == 10)))
assert(bool((sum(merge_kwargs({}, {}), 1, 2, 3) == 6)))
assert(bool((sum(merge_kwargs({}, {}), 1, 2) == 3)))
assert(bool((sum(merge_kwargs({second = 2}, {}), 1) == 3)))
assert(bool((sum(merge_kwargs({first = 1, second = 2}, {})) == 3)))
local function add(kvs, a, b, ...)
    a = get_posarg(kvs, 'a', a, nil, 'add')
    b = get_posarg(kvs, 'b', b, nil, 'add')
    local c = list {...}
    local d = get_kwonlyarg(kvs, 'd', nil, 'add')
    local e = get_kwonlyarg(kvs, 'e', nil, 'add')
    return (((a + b) + d) + e)
end
assert(bool((add(merge_kwargs({d = 1, e = 2}, {}), 1, 2, 3, 4) == 6)))
assert(bool((add(merge_kwargs({d = 1, e = 2}, {}), 1, 2) == 6)))
assert(bool((add(merge_kwargs({b = 2, d = 1, e = 2}, {}), 1) == 6)))
assert(bool((add(merge_kwargs({a = 1, b = 2, d = 1, e = 2}, {})) == 6)))
assert(bool((add(merge_kwargs({d = 1, e = 2, a = 1, b = 2}, {})) == 6)))
function add(kvs, a, b, ...)
    a = get_posarg(kvs, 'a', a, nil, 'add')
    b = get_posarg(kvs, 'b', b, nil, 'add')
    local c = list {...}
    local d = get_kwonlyarg(kvs, 'd', 0, 'add')
    local e = get_kwonlyarg(kvs, 'e', nil, 'add')
    return (((a + b) + d) + e)
end
assert(bool((add(merge_kwargs({d = 1, e = 2}, {}), 1, 2, 3, 4) == 6)))
assert(bool((add(merge_kwargs({e = 2}, {}), 1, 2, 3, 4) == 5)))
assert(bool((add(merge_kwargs({d = 1, e = 2}, {}), 1, 2) == 6)))
assert(bool((add(merge_kwargs({e = 2}, {}), 1, 2) == 5)))
assert(bool((add(merge_kwargs({b = 2, d = 1, e = 2}, {}), 1) == 6)))
assert(bool((add(merge_kwargs({b = 2, e = 2}, {}), 1) == 5)))
assert(bool((add(merge_kwargs({a = 1, b = 2, d = 1, e = 2}, {})) == 6)))
assert(bool((add(merge_kwargs({a = 1, b = 2, e = 2}, {})) == 5)))
assert(bool((add(merge_kwargs({d = 1, e = 2, a = 1, b = 2}, {})) == 6)))
assert(bool((add(merge_kwargs({e = 2, a = 1, b = 2}, {})) == 5)))
function add(kvs, a, b, ...)
    a = get_posarg(kvs, 'a', a, nil, 'add')
    b = get_posarg(kvs, 'b', b, nil, 'add')
    local c = list {...}
    local d = get_kwonlyarg(kvs, 'd', nil, 'add')
    local e = get_kwonlyarg(kvs, 'e', nil, 'add')
    local f = get_kwargs(kvs, {a = true, b = true, d = true, e = true})
    return (((((a + b) + d) + e) + f[_to_null("c")]) + f[_to_null("f")])
end
assert(bool((add(merge_kwargs({d = 1, e = 2, c = 3, f = 4}, {}), 1, 2, 3, 4) == 13)))
assert(bool((add(merge_kwargs({d = 1, e = 2, c = 3, f = 4}, {}), 1, 2) == 13)))
assert(bool((add(merge_kwargs({b = 2, d = 1, e = 2, c = 3, f = 4}, {}), 1) == 13)))
assert(bool((add(merge_kwargs({a = 1, b = 2, d = 1, e = 2, c = 3, f = 4}, {})) == 13)))
assert(bool((add(merge_kwargs({d = 1, e = 2, a = 1, b = 2, c = 3, f = 4}, {})) == 13)))
local function registerClickEvent(kvs, callback)
    callback = get_posarg(kvs, 'callback', callback, nil, 'registerClickEvent')
    --[[register a function as callback]]
    local event = dict {[_to_null("id")] = _to_null("click")}
    callback(merge_kwargs({}, {}), event)
end
local function callback(kvs, event)
    event = get_posarg(kvs, 'event', event, nil, 'callback')
    --[[simple callback]]
    assert(bool((event[_to_null("id")] == "click")))
end
registerClickEvent(merge_kwargs({}, {}), callback)
local function plus_factory(kvs, plus_num)
    plus_num = get_posarg(kvs, 'plus_num', plus_num, nil, 'plus_factory')
    --[[plus factory]]
    local function plus_func(kvs, n)
        n = get_posarg(kvs, 'n', n, nil, 'plus_func')
        --[[plus function]]
        return (n + plus_num)
    end
    return plus_func
end
local plus_10 = plus_factory(merge_kwargs({}, {}), 10)
assert(bool((plus_10(merge_kwargs({}, {}), 20) == 30)))
assert(bool((plus_10(merge_kwargs({}, {}), (-10)) == 0)))
local function clamp_wrapper(kvs, func)
    func = get_posarg(kvs, 'func', func, nil, 'clamp_wrapper')
    --[[a clamp decorator]]
    local function wrapper(kvs, ...)
        local args = list {...}
        local kwargs = get_kwargs(kvs, {})
        local res = func(merge_kwargs({}, kwargs), unpack(args))
        if bool((res < 0)) then
            res = 0
        end
        return res
    end
    return wrapper
end
local function minus(kvs, a, b, ...)
    a = get_posarg(kvs, 'a', a, nil, 'minus')
    b = get_posarg(kvs, 'b', b, 10, 'minus')
    local c = list {...}
    local d = get_kwonlyarg(kvs, 'd', 10, 'minus')
    local e = get_kwonlyarg(kvs, 'e', 1, 'minus')
    local f = get_kwargs(kvs, {a = true, b = true, d = true, e = true})
    return (a - d)
end
minus = clamp_wrapper({}, minus)
assert(bool((minus(merge_kwargs({}, {}), 10, 1) == 0)))
assert(bool((minus(merge_kwargs({}, {}), 10, 20) == 0)))
assert(bool((minus(merge_kwargs({}, {}), 1) == 0)))
assert(bool((minus(merge_kwargs({d = 0}, {}), 1) == 1)))
local function clamp_wrapper_with(kvs, left, right)
    left = get_posarg(kvs, 'left', left, nil, 'clamp_wrapper_with')
    right = get_posarg(kvs, 'right', right, nil, 'clamp_wrapper_with')
    --[[a clamp decorator with left and right limit]]
    local function clamp_wrapper(kvs, func)
        func = get_posarg(kvs, 'func', func, nil, 'clamp_wrapper')
        local function wrapper(kvs, ...)
            local args = list {...}
            local kwargs = get_kwargs(kvs, {})
            local res = func(merge_kwargs({}, kwargs), unpack(args))
            if bool((res < left)) then
                res = left
            elseif bool((res > right)) then
                res = right
            end
            return res
        end
        return wrapper
    end
    return clamp_wrapper
end
function sum(kvs, a, b, ...)
    a = get_posarg(kvs, 'a', a, nil, 'sum')
    b = get_posarg(kvs, 'b', b, nil, 'sum')
    local c = list {...}
    local d = get_kwonlyarg(kvs, 'd', nil, 'sum')
    local e = get_kwonlyarg(kvs, 'e', nil, 'sum')
    local f = get_kwargs(kvs, {a = true, b = true, d = true, e = true})
    local t = 0
    for _, i in c do
        t = (t + i)
        ::loop_label_2::
    end
    return ((((t + a) + b) + d) + e)
end
sum = clamp_wrapper_with(merge_kwargs({}, {}), 1, 10)({}, sum)
assert(bool((sum(merge_kwargs({d = 4, e = 5}, {}), 1, 2, 3) == 10)))
assert(bool((sum(merge_kwargs({d = (-3), e = (-4)}, {}), (-1), (-2)) == 1)))
assert(bool((sum(merge_kwargs({a = 1, b = (-1), c = 0, d = 1, e = (-10)}, {})) == 1)))
return {
    mul = mul,
    sub = sub,
    sum = sum,
    add = add,
    registerClickEvent = registerClickEvent,
    callback = callback,
    plus_factory = plus_factory,
    plus_10 = plus_10,
    clamp_wrapper = clamp_wrapper,
    minus = minus,
    clamp_wrapper_with = clamp_wrapper_with,
}
